src/controllers/update/payment.controller.ts

```diff
@@ -0,0 +1,31 @@
+import { Request, Response, NextFunction } from "express";
+import { createAndLogErrorRequest, logger } from "../../utils/logger";
+import * as config from "../../config";
+import { ApplicationData } from "model";
+import { getApplicationData } from "../../utils/application.data";
+import { OverseasEntityKey, PaymentKey } from "../../model/data.types.model";
+import { CreatePaymentRequest } from "@companieshouse/api-sdk-node/dist/services/payment";
+
+export const get = (req: Request, res: Response, next: NextFunction) => {
+  try {
+    const { status, state } = req.query;
+    const appData: ApplicationData = getApplicationData(req.session);
+    const savedPayment = appData[PaymentKey] || {} as CreatePaymentRequest;
+
+    logger.infoRequest(req, `Returned state: ${ state }, saved state: ${savedPayment.state}, with status: ${ status }`);
+    if (!savedPayment.state || savedPayment.state !== state) {
+      return next(createAndLogErrorRequest(req, `Rejecting update payment redirect, payment state does not match. Payment Request: ${ JSON.stringify(savedPayment)}`));
+    }
+
+    if (status === config.PAYMENT_PAID){
+      logger.debugRequest(req, `Overseas Entity id: ${ appData[OverseasEntityKey] }, Payment status: ${status}, Redirecting to: ${config.UPDATE_CONFIRMATION_URL}`);
+      return res.redirect(config.UPDATE_CONFIRMATION_URL);
+    } else {
+      logger.debugRequest(req, `Overseas Entity id: ${ appData[OverseasEntityKey] }, Payment status: ${status}, Redirecting to: ${config.UPDATE_CHECK_YOUR_ANSWERS_URL}`);
+      return res.redirect(config.UPDATE_CHECK_YOUR_ANSWERS_URL);
```

Comment:
We have an existing defect in registration (https://companieshouse.atlassian.net/browse/ROE-1891) for a similar line as if the payment fails or is cancelled then the user is able to change data from the check your answers screen which fails when our 'save and resume' feature is active.

Not sure if this is relevant for update but you might want to consider the behaviour/flow with your analysts and the UX team.

Reply:
Thanks for flagging that. We'd ideally like to do something better than just go back to the check your answers page and have a special retry payments page. This has been proposed but now we may have a good reason for doing it.